python/ray/air/tests/test_experiment_restore.py

```diff
@@ -0,0 +1,249 @@
+    """
+    This is an integration stress test for experiment restoration.
+
+
+    Test setup:
+
+    - For Tuner.restore:
+        - 8 trials, with a max of 2 running concurrently (--> 4 rounds of trials)
+        - Each iteration takes 0.5 seconds
+        - Each trial runs for 8 iterations --> 4 seconds
+        - Each round of 2 trials should take 4 seconds
+        - Without any interrupts/restoration:
+            - Minimum runtime: 4 rounds * 4 seconds / round = 16 seconds
+        - The test will stop the script with a SIGINT at a random time between
+        4-8 iterations each restore.
+
+    - For Trainer.restore:
+        - 1 trial with 4 workers
+        - Each iteration takes 0.5 seconds
+        - Runs for 32 iterations --> Minimum runtime = 16 seconds
+        - The test will stop the script with a SIGINT at a random time between
+        4-8 iterations after each restore.
+
+
+    Requirements:
+    - Req 1: Reasonable runtime
+        - The experiment should finish within 1.5 * 16 = 24 seconds.
+        - 1.5x is the passing threshold.
+        - 16 seconds is the minimum runtime.
+    - Req 2: Training progress persisted
+        - The experiment should progress monotonically.
+        (The training iteration shouldn't go backward at any point)
+        - Trials shouldn't start from scratch.
+    - Req 3: Searcher state saved/restored correctly
+    - Req 4: Callback state saved/restored correctly
+    """
+
+    np.random.seed(2023)
+
+    script_path = Path(__file__).parent / _RUN_SCRIPT_FILENAME
+
+    # Args to pass into the script as environment variables
+    exp_name = f"{runner_type}_restore_integration_test"
+    callback_dump_file = tmp_path / f"{runner_type}-callback_dump_file.json"
+    storage_path = tmp_path / "ray_results"
+    if storage_path.exists():
+        shutil.rmtree(storage_path)
+
+    csv_file = str(tmp_path / "dummy_data.csv")
+    dummy_df = pd.DataFrame({"x": np.arange(128), "y": 2 * np.arange(128)})
+    dummy_df.to_csv(csv_file)
+
+    run_started_marker = tmp_path / "run_started_marker"
+
+    time_per_iter_s = 0.5
+    max_concurrent = 2
+
+    if runner_type == "tuner":
+        iters_per_trial = 8
+        num_trials = 8
+    elif runner_type == "trainer":
+        iters_per_trial = 32
+        num_trials = 1
+
+    total_iters = iters_per_trial * num_trials
+
+    env = {
+        "RUNNER_TYPE": runner_type,
+        "STORAGE_PATH": str(storage_path),
+        "EXP_NAME": exp_name,
+        "CALLBACK_DUMP_FILE": str(callback_dump_file),
+        "RUN_STARTED_MARKER": str(run_started_marker),
+        "TIME_PER_ITER_S": str(time_per_iter_s),
+        "ITERATIONS_PER_TRIAL": str(iters_per_trial),
+        "NUM_TRIALS": str(num_trials),
+        "MAX_CONCURRENT_TRIALS": str(max_concurrent),
+        "CSV_DATA_FILE": csv_file,
+    }
+
+    # Pass criteria
+    no_interrupts_runtime = 16.0
+    passing_factor = 1.5
+    passing_runtime = no_interrupts_runtime * passing_factor
```

Comment:
I am a little worried about this hardcoded runtime, since tests can act quite differently on CI machines.
I hope this won't be flaky.

Reply:
@gjoliver I see, yeah I was a bit worried about this too. What about making the threshold much more lenient? Like 2x.

The `total_runtime` calculation is only adding up *actual training time*. On every run, it's calculated as the time between training started and when the run gets killed by interrupt. So, it's independent of extra time CI machines might take to initialize Trainable, and handle restoration, etc.